File: 백준/Gold/1167. 트리의 지름/트리의 지름.py
import sys
from collections import defaultdict

input = sys.stdin.readline
sys.setrecursionlimit(10**6)


def dfs(graph, start, visited):
    visited.add(start)
    dist_sum = 0
    last_node = start
    for node, distance in graph[start]:
        if node not in visited:
            next_node, total = dfs(graph, node, visited)
            total += distance
            if total > dist_sum:
                dist_sum = total
                last_node = next_node
    return last_node, dist_sum


def main():
    graph = defaultdict(list)
    # 1~V까지
    V = int(input())  # 2 ≤ V ≤ 100,000
    for i in range(1, V + 1):
        arr = list(map(int, input().split()))
        arr = arr[:-1]
        key = arr.pop(0)
        for k in range(0, len(arr), 2):
            graph[key].append((arr[k], arr[k + 1]))

    farthest, _ = dfs(graph, 1, set())
    print(dfs(graph, farthest, set())[1])


if __name__ == "__main__":
    main()

File: 백준/Gold/1167. 트리의 지름/트리의 지름.java
import java.io.*;
import java.util.*;

public class Main {
    public static int N;
    public static int[] dfs(int start, Map<Integer, List<int[]>> map, boolean[] visited){
        int dist_sum = 0;
        int last_node = start;
        visited[start] = true;
        List<int[]> neighbors = map.get(start);
        if(neighbors != null){
            for(int[] neighbor : neighbors){
                if(!visited[neighbor[0]]){
                    int[] next = dfs(neighbor[0], map, visited);
                    int tmp_cost = next[1] + neighbor[1];
                    if(tmp_cost > dist_sum){
                        dist_sum = tmp_cost;
                        last_node = next[0];
                    }
                }
            }
        }
        return new int[]{last_node, dist_sum};
    }
    public static void main(String[] args) throws IOException {
        // 입력을 빠르게 받기 위한 BufferedReader 사용
        BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(input.readLine());
        Map<Integer, List<int[]>> map = new HashMap<>();
        for(int i = 0; i<N; i++){
            String[] data = input.readLine().split(" ");
            int start = Integer.parseInt(data[0]);
            for(int k = 1; k < data.length-1; k+=2){
                int end = Integer.parseInt(data[k]);
                int cost = Integer.parseInt(data[k+1]);
                map.computeIfAbsent(start, arr -> new ArrayList<>()).add(new int[]{end, cost});
            }
        }
        int[] first = dfs(1, map, new boolean[N+1]);
        System.out.println(dfs(first[0], map, new boolean[N + 1])[1]);
    }
}